backend/src/apiserver/storage/job_store.go

```diff
@@ -40,41 +40,80 @@ type JobStore struct {
 	time                   util.TimeInterface
 }
 
+// Runs two SQL queries in a transaction to return a list of matching jobs, as well as their
+// total_size. The total_size does not reflect the page size.
 func (s *JobStore) ListJobs(
-	filterContext *common.FilterContext, opts *list.Options) ([]*model.Job, string, error) {
-	errorF := func(err error) ([]*model.Job, string, error) {
-		return nil, "", util.NewInternalServerError(err, "Failed to list jobs: %v", err)
+	filterContext *common.FilterContext, opts *list.Options) ([]*model.Job, int, string, error) {
+	errorF := func(err error) ([]*model.Job, int, string, error) {
+		return nil, 0, "", util.NewInternalServerError(err, "Failed to list jobs: %v", err)
 	}
 
-	sqlBuilder := s.selectJob()
-	// Add filter condition
-	sqlBuilder, err := s.toFilteredQuery(sqlBuilder, filterContext)
+	// SQL for getting the filtered and paginated rows
+	rowsQuery, err := s.toFilteredQuery(s.selectJob(), filterContext)
+	if err != nil {
+		return errorF(err)
+	}
+	rowsSql, rowsArgs, err := opts.AddPaginationToSelect(rowsQuery).ToSql()
 	if err != nil {
 		return errorF(err)
 	}
 
-	sql, args, err := opts.AddToSelect(sqlBuilder).ToSql()
+	// SQL for getting total count of the filtered rows
+	// This first query performs a filtered selection over the jobs table, it should match
+	// the number of rows returned by the filtered query built above for the actual rows,
+	// but it tries to optimize out the left join, which isn't needed here for the count.
+	countQuery, err := s.toFilteredQuery(sq.Select("*").From("jobs"), filterContext)
+	if err != nil {
+		return errorF(err)
+	}
+	countSql, countArgs, err := sq.Select("count(*)").FromSelect(countQuery, "rows").ToSql()
+	// countSql, countArgs, err := countQuery.ToSql()
 	if err != nil {
 		return errorF(err)
 	}
 
-	rows, err := s.db.Query(sql, args...)
+	// Use a transaction to make sure we're returning the total_size of the same rows queried
+	tx, err := s.db.Begin()
 	if err != nil {
 		return errorF(err)
 	}
-	defer rows.Close()
 
+	rows, err := tx.Query(rowsSql, rowsArgs...)
+	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
 	jobs, err := s.scanRows(rows)
 	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
+	rows.Close()
+
+	countRow, err := tx.Query(countSql, countArgs...)
+	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
+	total_size, err := ScanRowToTotalSize(countRow)
+	if err != nil {
+		tx.Rollback()
+		return errorF(err)
+	}
+	countRow.Close()
+
+	tx.Commit()
+	if err != nil {
```

Comment:
Please have a look at the sample below. (from https://golang.org/pkg/database/sql/)

It explains the proper way to handle failed calls to rollback/commit. A failed commit or a failed rollback should be logged. Why are you calling rollback after the failed commit? 

I would suggest to 
1) log errors from the call to commit
2) log errors from the call to rollback
3) do not call rollback in case of failed commit. 

package main

import (
	"context"
	"database/sql"
	"log"
)

var ctx = context.Background()
var db *sql.DB

func main() {
	tx, err := db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable})
	if err != nil {
		log.Fatal(err)
	}
	id := 53
	_, err = tx.ExecContext(ctx, "UPDATE drivers SET status = ? WHERE id = ?", "assigned", id)
	if err != nil {
		if rollbackErr := tx.Rollback(); rollbackErr != nil {
			log.Printf("Could not roll back: %v\n", rollbackErr)
		}
		log.Fatal(err)
	}
	_, err = tx.ExecContext(ctx, "UPDATE pickups SET driver_id = $1", id)
	if err != nil {
		if rollbackErr := tx.Rollback(); rollbackErr != nil {
			log.Printf("Could not roll back: %v\n", rollbackErr)
		}
		log.Fatal(err)
	}
	if err := tx.Commit(); err != nil {
		log.Fatal(err)
	}
}

Reply:
Please review the other instances as well.